pkg/checkpoint/secret.go

```
@@ -50,14 +53,71 @@ func (c *checkpointer) checkpointSecret(namespace, podName, secretName string, u
 
 	// TODO(aaron): No need to store if already exists
 	for f, d := range secret.Data {
-		if err := writeAndAtomicRename(filepath.Join(basePath, f), d, uid, gid, 0600); err != nil {
+		if err := writeAndAtomicRename(filepath.Join(basePath, f), d, uid, gid, defaultSecretMode); err != nil {
 			return "", fmt.Errorf("failed to write secret %s: %v", secret.Name, err)
 		}
 	}
 
 	return basePath, nil
 }
 
+// checkpointSecretProjection will locally store secret data from volumes with ProjectedVolumeSources.
+// The path to the secret data becomes: checkpointSecretPath/namespace/podname/volumename/secret.file
+// Where each "secret.file" is a path from the SecretProjection.KeyToPath field.
+func (c *checkpointer) checkpointSecretProjection(namespace, podName, volumeName string, secretProjection *v1.SecretProjection, uid, gid int) error {
+
+	secretName := secretProjection.Name
+	isOptional := secretProjection.Optional != nil && *secretProjection.Optional
+
+	secret, err := c.apiserver.Core().Secrets(namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil && isOptional {
+		return nil
```

Same nit as above: potentially hiding errors we should surface